lesson-04-memory-management/mm/xxx.c

```diff
@@ -5,21 +5,142 @@
 #include <linux/pci.h>
 #include <linux/version.h>
 #include <linux/init.h>
+#include <linux/gfp.h>
+#include <linux/sysfs.h>
 
-#define LEN_MSG 160
-static char buf_msg[ LEN_MSG + 1 ] = "Hello from module!\n";
+static size_t g_buf_size = 0;
+static char* g_buf_msg   = 0;
 
+#define MEM_CONFIG_KMALLOC 0
+#define MEM_CONFIG_KMCACHE 1
+
+#define CACHE_NAME "xxx_cache"
+#define CACHE_SIZE 160
+static struct kmem_cache* g_cache = NULL;
+
+static void cache_constructor( void* p )
+{
+   printk( "%s constructs %p(%u)", THIS_MODULE->name, p, (p ? *(char*)p : 0) );
+}
+
+static int g_mem_config = MEM_CONFIG_KMALLOC;
+module_param_named( mem_config, g_mem_config, int, 0 );
+
+
+static void initialize_memory( void )
+{
+   if (g_mem_config != MEM_CONFIG_KMCACHE)
+      return;
+
+   g_cache = kmem_cache_create( CACHE_NAME, CACHE_SIZE, 0/*SLAB_HWCACHE_ALIGN*/,
+      0/*SLAB_DEBUG_INITIAL*/, &cache_constructor );
+   printk( "%s %s cache: %p", THIS_MODULE->name, __FUNCTION__, g_cache );
+}
+
+static void finalize_memory( void )
+{
+   if (g_mem_config != MEM_CONFIG_KMCACHE)
+      return;
+
+   if (g_cache)
+      kmem_cache_destroy( g_cache );
+   g_cache = NULL;
+}
+
+static void* allocate_memory( size_t* count )
+{
+   void* result = NULL;
+   if (!count)
+      return result;
+
+   if (g_mem_config == MEM_CONFIG_KMALLOC)
+   {
+      result = kmalloc( *count, GFP_ATOMIC );
+   }
+   else if (g_mem_config == MEM_CONFIG_KMCACHE)
+   {
```

Comment:
Such implementation with `kmem_cache` actually doesn't make much sense - for data which exceed buffer size you just allocate new blocks in a loop and throw away previous one - to implement resizing you'll need to implement some linked list on these blocks.

Reply:
not perfect, but done